Composer/packages/tools/language-servers/language-generation/src/LGServer.ts

```diff
@@ -0,0 +1,355 @@
+      return new Promise<string>((resolve, reject) => {
+        // eslint-disable-next-line security/detect-non-literal-fs-filename
+        fs.readFile(uri.fsPath, 'UTF-8', (err, result) => {
+          err ? reject('') : resolve(result.toString());
+        });
+      });
+    }
+    try {
+      const response = await xhr({ url, followRedirects: 5 });
+      return response.responseText;
+    } catch (error) {
+      return Promise.reject(error.responseText || getErrorStatusDescription(error.status) || error.toString());
+    }
+  }
+
+  private removeParamFormat(params: string): string {
+    const paramArr: string[] = params.split(',');
+    const resultArr: string[] = [];
+    paramArr.forEach(element => {
+      resultArr.push(element.trim().split(':')[0]);
+    });
+    return resultArr.join(' ,');
+  }
+
+  private matchedStates(params: TextDocumentPositionParams): { matched: boolean; state: string } | null {
+    const state: string[] = [];
+    const document = this.documents.get(params.textDocument.uri);
+    if (!document) return null;
+    const position = params.position;
+    const range = Range.create(position.line, 0, position.line, position.character);
+    const lineContent = document.getText(range);
+    let flag = false;
+    let finalState = '';
+    if (!lineContent.trim().startsWith('-')) {
+      return { matched: flag, state: finalState };
+    }
+
+    //if the line starts with '-', will try to match
+    flag = true;
+
+    //initialize the root state to plaintext
+    state.push('PlainText');
+
+    // find out the context state of current cursor, offer precise suggestion and completion etc.
+    /**
+     * - Hi, @{name}, what's the meaning of 'state'
+     * - Hi---------, @{name}--------, what-------' ------s the meaning of "state"
+     * - <plaintext>, @{<expression>}, <plaintext><single><plaintext>------<double>
+     */
+    let i = 0;
+    while (i < lineContent.length) {
+      const char = lineContent.charAt(i);
+      if (char === `'`) {
+        if (state[state.length - 1] === 'expression' || state[state.length - 1] === 'double') {
+          state.push('single');
+        } else {
+          state.pop();
+        }
+      }
+
+      if (char === `"`) {
+        if (state[state.length - 1] === 'expression' || state[state.length - 1] === 'single') {
+          state.push('double');
+        } else {
+          state.pop();
+        }
+      }
+
+      if (char === '{' && i >= 1 && state[state.length - 1] !== 'single' && state[state.length - 1] !== 'double') {
+        if (lineContent.charAt(i - 1) === '@') {
+          state.push('expression');
+        }
+      }
+
+      if (char === '}' && state[state.length - 1] === 'expression') {
+        state.pop();
+      }
+      i = i + 1;
+    }
+    finalState = state[state.length - 1];
+    return { matched: flag, state: finalState };
+  }
+
+  protected completion(params: TextDocumentPositionParams): Thenable<CompletionList | null> {
```

Comment:
Do this function need to return Promise?

Reply:
same as hover, see above comments